README.md

```diff
@@ -5,14 +5,25 @@
 
 ## Routes
 
-### `GET /retrieval`
+### `PUT /retrieval`
+
+Start a new retrieval.
 
 Response:
 
-```json
-{}
+```typescript
+{
+  id: String,
+  cid: String,
+  providerAddress: String,
+  protocol: 'graphsync'|'bitswap'
+}
 ```
 
+### `POST /retrieval/:id`
```

Comment:
See my comment above.

I am not sure if the job id should be in the endpoint path. My personal preference is to put it into the request payload.

In REST APIs, ids in paths are usually used to identify the resource we want to interact with.

When creating a new retrieval report, the path should contain the report id, not the job id.

Maybe I have a different mental model of the data schema you want to implement?

Reply:
I think the difference is that you're thinking about `retrieval job` and `retrieval report`, while I'm only thinking about one object `retrieval`, which gets updated as we go.

I like having just one object for a retrieval. You don't need to join when querying, and also job+report have a 1:1 relationship, so we don't need flexibility gained by decoupling the data.

Reply:
Fair enough. 

It's possible for multiple clients to submit retrieval data for the same job (retrieval). I think we should record such attempts and use the data to detect fraudulent Station operators. I guess your schema does not prevent us from doing so - the first "submit" request will update the database, and subsequent requests will detect that the report was already submitted and record the fraudulent attempt elsewhere.

In that case I am proposing:

(1)
`POST /retrievals` to create a new retrieval


(2)
`PATCH /retrievals/:id` to report the retrieval outcome. Why `PATCH`? Because `PUT` is usually used to update all data fields of a resource. 

https://restfulapi.net/http-methods/#patch
> HTTP PATCH requests are to make a partial update on a resource.
>
> If you see PUT requests modify a resource entity too. So to make it more precise – the PATCH method is the correct choice for partially updating an existing resource, and you should only use PUT if you’re replacing a resource in its entirety.

TBH, I think it's better to abandon the REST design philosophy for this second endpoint and use the RPC style instead:

`POST /submit-retrieval-report` with all data provided in the request body.

Reply:
Aaaah right, POST and PATCH is the right pair 👍 

> TBH, I think it's better to abandon the REST design philosophy for this second endpoint and use the RPC style instead:

Why?

Reply:
> Aaaah right, POST and PATCH is the right pair 👍

👌🏻 

> > TBH, I think it's better to abandon the REST design philosophy for this second endpoint and use the RPC style instead:
> 
> Why?

IMO, we don't want clients (the Spark module) to change a subset of `retrieval` properties arbitrarily, which is my understanding of PATCH semantics. Instead, we want them to submit a retrieval result with metrics, which is a strictly defined RPC-like operation with a list of required request fields.

Anyhow, I don't hold this opinion too strongly. I can live with `PATCH` too.

Reply:
Implementing PATCH doesn't mean you're allowed to change all fields, it can still only be just the ones we need.